### PR TITLE
Avoid early error for transistive constraints

### PR DESCRIPTION
**Pull request by charliermarsh**
## Summary

If a constraint causes some requirement in a dependency to yield the empty set, we should avoid early termination. (I believe we do this right now because the error messages are better.) This is different than if you have (e.g.) two direct requirements with conflicting versions, which _should_ error immediately.

Closes https://github.com/astral-sh/uv/issues/1522.

### CODE REVIEW

File: crates/uv/tests/pip_compile.rs
Comment (charliermarsh) [inline crates/uv/tests/pip_compile.rs:3595]: I need to improve this case before merging.

**Comment by charliermarsh** [comment]
@zanieb - I think this is the thing that would've been solved by that PR you had at one point to allow returning multiple versions (i.e., to avoid this eager merging).

**Comment by zanieb** [comment]
I'm not sure which one you're referring to anymore :)

**Comment by charliermarsh** [comment]
@zanieb - Found it! https://github.com/astral-sh/uv/pull/383

**Comment by zanieb** [comment]
Oh gosh we probably still want that don't we... 😭

**Comment by charliermarsh** [comment]
Yeah I think it would indirectly solve this problem

**Comment by charliermarsh** [comment]
(Like in theory it would make this PR redundant and avoid the bad error message I commented on in the PR.)

**Comment by zanieb** [comment]
I think the main blocker there was the URL dependency stuff and I think you had a better solution for that than me? Maybe you should take a look at it?

**Comment by charliermarsh** [comment]
@zanieb - Yeah I can take this on. I have to prioritize some other things over it first. Do you think this is worth merging as-is even with the bad error message? Defer to you on that.

**Review by zanieb — APPROVED** [review]
I'll take the incremental improvement.

Is there a test case covering a successful resolution for the bug? It looks like the new test case does not succeed?

**Comment by charliermarsh** [comment]
Closing in favor of https://github.com/astral-sh/uv/pull/1796.

**Pull request closed without merging**
